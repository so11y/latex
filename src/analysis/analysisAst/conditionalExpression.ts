import { Node, ConditionalExpression, BinaryExpression } from "estree";
import { AstType, ValidateSchemaGuardMate, cratedNotThrough } from "../types";

import { ValidateSchemaBase } from "../types";
import { ErrorMessage, formatterError } from "../helper/errorMessage";
import {
  LogicalOperators,
  isLogicalOperators,
  isSafeOperators,
} from "../helper/defineOperators";

function trueAndFalseResult(
  node: Node,
  message: string
): true | ValidateSchemaGuardMate {
  if (node.type !== AstType.CallExpression) {
    return cratedNotThrough(node, message);
  }
  return true;
}

//LogicalExpression | BinaryExpression | CallExpression
export function validateIsLogicalNode(
  node: Node
): true | ValidateSchemaGuardMate {
  if (node.type === AstType.ConditionalExpression) {
    return cratedNotThrough(
      node,
      ErrorMessage.ConditionalExpression.NotNestConditionalExpression
    );
  }

  const notLogicOrBinary = [
    AstType.BinaryExpression,
    AstType.LogicalExpression,
  ].every((type) => node.type !== type);

  if (notLogicOrBinary) {
    return cratedNotThrough(
      node,
      formatterError`${
        ErrorMessage.ConditionalExpression.OnlyLogical
      } ${LogicalOperators.join(" ")}`
    );
  }

  if (AstType.BinaryExpression === node.type) {
    const { through, message } = isSafeOperators(
      (node as BinaryExpression).operator
    );
    if (through === false) {
      return cratedNotThrough(node, message!);
    }
    //如果是二元表达式那么它的操作符必须是逻辑表达式
    if (isLogicalOperators((node as BinaryExpression).operator) === false) {
      return cratedNotThrough(
        node,
        ErrorMessage.ConditionalExpression.OnlyLogical
      );
    }
  }

  return true;
}

export const ConditionalExpressionSchema: ValidateSchemaBase = {
  type: "ConditionalExpression",
  validate(node: ConditionalExpression) {
    const { test, consequent, alternate } = node;

    if (![test, consequent, alternate].every(Boolean)) {
      return cratedNotThrough(node, "需要三个参数");
    }

    const errorNodes = [
      validateIsLogicalNode(test),
      trueAndFalseResult(
        consequent,
        ErrorMessage.ConditionalExpression.OnlyTrueCall
      ),
      trueAndFalseResult(
        alternate,
        ErrorMessage.ConditionalExpression.OnlyFalseCall
      ),
    ]
      .map((result) => {
        return result === true
          ? null
          : cratedNotThrough(result.node, result.message!);
      })
      .filter(Boolean) as Array<ValidateSchemaGuardMate<Node>>;

    if (errorNodes.length) {
      return cratedNotThrough(node, null, errorNodes);
    }
  },
};
