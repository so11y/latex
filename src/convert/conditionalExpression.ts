import { decode, encode, newId } from "./index";
import { StoreContainProvide } from "../components/storeContain";
import { ContainLatexComponentStore } from "../components/types";
import { Ast } from "./ast";
import { buildConfig } from "./helper";

export class ConditionalExpression extends Ast {
  static type = "ConditionalExpression";
  encode(v: ContainLatexComponentStore, s: StoreContainProvide) {
    const [test, consequent, alternate] = v.children.value;
    return {
      type: ConditionalExpression.type,
      config: buildConfig(v),
      test: encode(s.getComponent(test.id), s),
      consequent: encode(s.getComponent(consequent.id), s),
      alternate: encode(s.getComponent(alternate.id), s),
    };
  }
  async decodeTemplate(v: Record<string, any>, s: StoreContainProvide) {
    const { test, consequent, alternate } = v as ReturnType<
      ConditionalExpression["encode"]
    >;
    await decode(test, s);
    await decode(consequent, s);
    await decode(alternate, s);
  }

  newIdTemplate(
    v: Record<string, any>,
    s: StoreContainProvide,
    parentId: string
  ) {
    const { config } = v;
    config.parentId = parentId;
    config.id = s.getIncreaseUid();

    const { test, consequent, alternate } = v as ReturnType<
      ConditionalExpression["encode"]
    >;
    newId(test, s, config.id);
    newId(consequent, s, config.id);
    newId(alternate, s, config.id);
  }
}
