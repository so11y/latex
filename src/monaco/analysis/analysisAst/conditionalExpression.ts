import { Node, ConditionalExpression, BinaryExpression } from "estree";
import { AstType, ValidateSchemaGuardMate, cratedNotThrough } from "../types";
import {
  LogicalOperators,
  isLogicalOperators,
  isSafeOperators,
} from "../../util";

import { ValidateSchemaBase } from "../types";

export type ConditionalExpressionSchemeType = Omit<
  ValidateSchemaBase,
  "type"
> & {
  type: "ConditionalExpression";
};

const ErrorMessage: Array<string> = [
  `只能是使用三元和逻辑表达式符号 ${LogicalOperators.join(" ")}`,
  "真结果需要是函数调用并且不能再返回逻辑表达式了",
  "假结果需要是函数调用并且不能再返回逻辑表达式了",
  `不支持在直接嵌套Conditional函数和三元表达式,但是可以在搭配逻辑表达式使用`,
];

function trueAndFalseResult(
  node: Node,
  message: string
): true | ValidateSchemaGuardMate {
  if (node.type !== AstType.CallExpression) {
    return {
      through: false,
      node,
      message,
    };
  }
  return true;
}

//LogicalExpression | BinaryExpression | CallExpression
export function validateIsLogicalNode(
  node: Node
): true | ValidateSchemaGuardMate {
  const notLogicOrBinaryOrConditional = [
    AstType.BinaryExpression,
    AstType.LogicalExpression,
    AstType.ConditionalExpression,
  ].every((type) => node.type !== type);

  if (notLogicOrBinaryOrConditional) {
    return {
      through: false,
      node,
      message: ErrorMessage[0],
    };
  }

  if (AstType.BinaryExpression === node.type) {
    const { through, message } = isSafeOperators(
      (node as BinaryExpression).operator
    );
    if (through === false) {
      return {
        through: false,
        node,
        message: message!,
      };
    }
    //如果是二元表达式那么它的操作符必须是逻辑表达式
    if (isLogicalOperators((node as BinaryExpression).operator) === false) {
      return {
        through: false,
        node,
        message: ErrorMessage[0],
      };
    }
  }

  return true;
}

export const ConditionalExpressionSchema: ConditionalExpressionSchemeType = {
  type: "ConditionalExpression",
  validate(node: ConditionalExpression) {
    const { test, consequent, alternate } = node;

    if (![test, consequent, alternate].every(Boolean)) {
      return cratedNotThrough(node, "需要三个参数");
    }

    const errorNodes = [
      validateIsLogicalNode(test),
      trueAndFalseResult(consequent, ErrorMessage[1]),
      trueAndFalseResult(alternate, ErrorMessage[2]),
    ]
      .map((result) => {
        return result === true
          ? null
          : cratedNotThrough(result.node, result.message!);
      })
      .filter(Boolean) as Array<ValidateSchemaGuardMate<Node>>;

    if (errorNodes.length) {
      return cratedNotThrough(node, null, errorNodes);
    }
  },
};
