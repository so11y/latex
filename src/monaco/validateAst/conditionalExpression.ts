import {
  CallExpression,
  LogicalExpression,
  BinaryExpression,
  Node,
} from "acorn";
import { AstType, ValidateSchemaGuardMate, cratedNotThrough } from "./types";
import { LogicalOperators, isLogicalOperators, isSafeOperators } from "../util";
import { LatexValidateConfig, LatexCallConfig } from "./latexConfig";

import { LatexNames } from "./latexConfig";
import { ValidateSchemaBase } from "./types";
import { ConditionalExpression } from "acorn";

export type ConditionalExpressionSchemeType = Omit<
  ValidateSchemaBase,
  "type"
> & {
  type: "ConditionalExpression";
};

const ErrorMessage: Array<string> = [
  `只能是使用三元和逻辑表达式符号 ${LogicalOperators.join(" ")}`,
  "真结果需要是函数调用并且不能再返回逻辑表达式了",
  "真结果需要是函数调用并且不能再返回逻辑表达式了",
  `不支持在直接嵌套Conditional函数和三元表达式,但是可以在搭配逻辑表达式使用`,
];

function trueAndFalseResult(
  node: Node,
  message: string
): true | ValidateSchemaGuardMate {
  if (node.type !== AstType.CallExpression) {
    return {
      through: false,
      node,
      message,
    };
  }
  return true;
}

//LogicalExpression | BinaryExpression | CallExpression
export function validateIsLogicalNode(
  node: Node
): true | ValidateSchemaGuardMate {
  const notLogicOrBinaryOrConditional = [
    AstType.BinaryExpression,
    AstType.LogicalExpression,
    AstType.ConditionalExpression,
  ].every((type) => node.type !== type);

  if (notLogicOrBinaryOrConditional) {
    return {
      through: false,
      node,
      message: ErrorMessage[0],
    };
  }

  if (AstType.BinaryExpression === node.type) {
    const { through, message } = isSafeOperators(
      (node as BinaryExpression).operator
    );
    if (through === false) {
      return {
        through: false,
        node,
        message: message!,
      };
    }
    //如果是二元表达式那么它的操作符必须是逻辑表达式
    if (isLogicalOperators((node as BinaryExpression).operator) === false) {
      return {
        through: false,
        node,
        message: ErrorMessage[0],
      };
    }
  }

  return true;
}

//

export const ConditionalExpressionSchema: ConditionalExpressionSchemeType = {
  type: "ConditionalExpression",
  validate(node: ConditionalExpression) {
    const { test, consequent, alternate } = node;

    const errorNodes = [
      validateIsLogicalNode(test),
      trueAndFalseResult(consequent, ErrorMessage[1]),
      trueAndFalseResult(alternate, ErrorMessage[2]),
    ]
      .map((result) => {
        return result === true
          ? null
          : cratedNotThrough(result.node, result.message!);
      })
      .filter(Boolean) as Array<ValidateSchemaGuardMate<Node>>;

    if (errorNodes.length) {
      return cratedNotThrough(node, null, errorNodes);
    }
  },
};
