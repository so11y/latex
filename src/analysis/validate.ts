import { BinaryExpressionDefine } from "./analysisAst/binaryExpression";
import { CallExpressionDefine } from "./analysisAst/callExpression";
import * as monaco from "monaco-editor";
import { ConditionalExpressionDefine } from "./analysisAst/conditionalExpression";
import { ExpressionStatementDefine } from "./analysisAst/expressionStatement";
import { IdentifierDefine } from "./analysisAst/identifier";
import { LiteralDefine } from "./analysisAst/literal";
import { LogicalExpressionDefine } from "./analysisAst/logicalExpression";
import { NumberLiteralDefine } from "./analysisAst/numberLiteral";
import { ProgramDefine } from "./analysisAst/program";
import { AstType, ValidateDefineBase, ValidateGuardFalseMate } from "./types";
import { parse } from "./parse";
import { Program } from "acorn";
import { CallExpression, Node } from "estree";
import {
  cratedFakeNodeError,
  extractTokenAndNumbers,
  nomadizeMarkers,
} from "./util/functional";
import { LatexNames, macroLatexCallConfig } from "./helper/latexConfig";
import { walk } from "estree-walker";

export class Latex {
  static instance: Latex;

  static getInstance() {
    //测试的情况下，每次都会创建新的实例
    if (!Latex.instance || __Test__) {
      Latex.instance = new Latex();
    }
    return Latex.instance;
  }

  declare syntax: {
    mappings: Map<string, ValidateDefineBase>;
    typeKeys: string[];
  };

  ast: Program | null = null;

  hintsCallExpressionNodes: Array<CallExpression> = [];

  constructor() {
    this.syntax = buildTypeNames([
      CallExpressionDefine,
      ExpressionStatementDefine,
      IdentifierDefine,
      LiteralDefine,
      ProgramDefine,
      BinaryExpressionDefine,
      LogicalExpressionDefine,
      ConditionalExpressionDefine,
      NumberLiteralDefine,
    ]);
  }

  validate(value: string, ast?: Program | null) {
    const diagnosisNodes: Array<ValidateGuardFalseMate> = [];
    try {
      this.ast =
        ast ||
        parse.parse(value, {
          ecmaVersion: "latest",
          sourceType: "script",
          locations: true,
        });

      this.walk(this.ast as Node, diagnosisNodes);
    } catch (error) {
      if (diagnosisNodes.length === 0) {
        const position = extractTokenAndNumbers((error as Error).message);
        if (position) {
          diagnosisNodes.push(cratedFakeNodeError(position));
        }
      }
      console.log("ignore parse error");
    }
    if (diagnosisNodes.length !== 0) {
      this.ast = null;
    }
    return {
      diagnosisNodes,
      ast: this.ast,
    };
  }

  walk(ast: Node, diagnosisNodes: Array<ValidateGuardFalseMate>) {
    const _this = this;
    walk(ast as any, {
      enter(node, parent, prop, index) {
        _this.isNeedInlayHints(node);
        const schemas = _this.syntax.mappings.get(node.type)!;
        const validate = schemas.validate(node, parent, prop, index);
        if (validate.through === false) {
          diagnosisNodes.push(validate);
          this.skip();
          return;
        }
        if (validate.through === true && validate.eatKeys) {
          validate.eatKeys.forEach((key) => {
            const nestAst = (node as any)[key];
            if (Array.isArray(nestAst)) {
              (nestAst as Array<Node>).forEach((node) => {
                node.isEat = true;
              });
            } else {
              (nestAst as Node).isEat = true;
            }
          });
        }
      },
      leave(node) {
        if (!_this.syntax.typeKeys.includes(node.type) && !!node.isEat) {
          diagnosisNodes.push({
            node: node as Node,
            through: false,
          });
          this.skip();
          return;
        }
      },
    });
  }

  isNeedInlayHints(node: Node) {
    //提供編輯器類型的嵌入提示
    this.hintsCallExpressionNodes = [];
    if (
      node.type === AstType.CallExpression &&
      node.callee.type === AstType.Identifier
    ) {
      if (LatexNames.includes(node.callee.name)) {
        this.hintsCallExpressionNodes.push(node as CallExpression);
      }
    }
    if (node.type === AstType.ConditionalExpression) {
      this.hintsCallExpressionNodes.push({
        type: AstType.CallExpression,
        loc: node.loc,
        optional: false,
        callee: {
          type: AstType.Identifier,
          loc: null,
          name: macroLatexCallConfig.Conditional.name,
        },
        arguments: [node.test, node.consequent, node.alternate].filter(Boolean),
      });
    }
  }

  getMarkers(value: string, ast?: Program | null) {
    let { diagnosisNodes } = this.validate(value, ast);
    const markers: monaco.editor.IMarkerData[] = nomadizeMarkers(
      diagnosisNodes,
      []
    );
    return {
      ast: this.ast,
      markers,
    };
  }
}

function buildTypeNames(
  types: Array<Array<ValidateDefineBase> | ValidateDefineBase>
) {
  const mappings: Map<string, ValidateDefineBase> = new Map();

  for (const value of types) {
    let current: ValidateDefineBase = value as ValidateDefineBase;
    mappings.set(current.type, current);
  }

  return {
    mappings,
    typeKeys: Array.from(mappings.keys()),
  };
}
