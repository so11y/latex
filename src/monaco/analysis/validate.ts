import { CallExpressionSchema } from "./analysisAst/callExpression";
import { ExpressionStatementSchema } from "./analysisAst/expressionStatement";
import { IdentifierSchema } from "./analysisAst/identifier";
import { LiteralSchema } from "./analysisAst/literal";
import { ProgramSchema } from "./analysisAst/program";
import { BinaryExpressionSchema } from "./analysisAst/binaryExpression";
import { LogicalExpressionSchema } from "./analysisAst/logicalExpression";
import { ConditionalExpressionSchema } from "./analysisAst/conditionalExpression";
import { Node, CallExpression } from "estree";
import { walk } from "estree-walker";
import { AstType, ValidateSchemaBase, ValidateSchemaGuardMate } from "./types";
import { LatexNames, macroLatexCallConfig } from "./latexConfig";

function buildTypeNames(
  types: Array<Array<ValidateSchemaBase> | ValidateSchemaBase>
) {
  const mappings: Map<string, ValidateSchemaBase> = new Map();

  for (const value of types) {
    let current: ValidateSchemaBase = value as ValidateSchemaBase;
    mappings.set(current.type, current);
  }

  return {
    mappings,
    typeKeys: Array.from(mappings.keys()),
  };
}

const { typeKeys, mappings } = buildTypeNames([
  CallExpressionSchema,
  ExpressionStatementSchema,
  IdentifierSchema,
  LiteralSchema,
  ProgramSchema,
  BinaryExpressionSchema,
  LogicalExpressionSchema,
  ConditionalExpressionSchema,
]);

let hintsCallExpressionNodes: Array<CallExpression> = [];
export function getCallExpressionNodes() {
  return hintsCallExpressionNodes;
}

function isNeedInlayHints(node: Node) {
  //提供編輯器類型的嵌入提示
  if (
    node.type === AstType.CallExpression &&
    node.callee.type === AstType.Identifier
  ) {
    if (LatexNames.includes(node.callee.name)) {
      hintsCallExpressionNodes.push(node as CallExpression);
    }
  }
  if (node.type === AstType.ConditionalExpression) {
    hintsCallExpressionNodes.push({
      type: AstType.CallExpression,
      loc: node.loc,
      optional: false,
      callee: {
        type: AstType.Identifier,
        loc: null,
        name: macroLatexCallConfig.Conditional.name,
      },
      arguments: [node.test, node.consequent, node.alternate].filter(Boolean),
    });
  }
}

export function validateWalk(
  ast: Node,
  diagnosisNodes: Array<ValidateSchemaGuardMate>
) {
  hintsCallExpressionNodes = [];
  walk(ast as any, {
    enter(node, parent, prop, index) {
      isNeedInlayHints(node);
      if (!typeKeys.includes(node.type)) {
        diagnosisNodes.push({
          node: node as Node,
          through: false,
        });
        this.skip();
        return;
      }
      const schemas = mappings.get(node.type)!;
      const validate = schemas.validate(node, parent, prop, index);
      if (validate && validate.through === false) {
        diagnosisNodes.push(validate);
        this.skip();
      }
    },
  });
}
