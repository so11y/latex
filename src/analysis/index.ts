import * as monaco from "monaco-editor";
import { Latex } from "./validate";

export function handleValidate(value: string, model: monaco.editor.ITextModel) {
  let { ast, markers } = Latex.getInstance().getMarkers(value);
  monaco.editor.setModelMarkers(model, "owner", markers);
  return {
    ast,
  };
}
