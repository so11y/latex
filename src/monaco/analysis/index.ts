import * as monaco from "monaco-editor";
import { Program } from "acorn";
import { extractTokenAndNumbers } from "./util";
import { ValidateSchemaGuardMate } from "./types";
import { validateWalk } from "./validate";
import { Node } from "estree";
import { parse } from "./parse";

export function handleValidate(value: string, model: monaco.editor.ITextModel) {
  let { ast, diagnosisNodes } = validate(value);
  const markers: monaco.editor.IMarkerData[] = [];
  nomadizeMarkers(diagnosisNodes);
  function nomadizeMarkers(nodes: ValidateSchemaGuardMate<Node>[]) {
    for (const node of nodes) {
      if (node.errorNodes?.length) {
        nomadizeMarkers(node.errorNodes);
      } else {
        const { loc } = node.node;
        markers.push({
          message: node.message ?? "未知字符",
          severity: node.severity ?? monaco.MarkerSeverity.Error,
          startLineNumber: loc!.start.line,
          startColumn: loc!.start.column + 1,
          endLineNumber: loc!.end.line,
          endColumn: loc!.end.column + 1,
        });
      }
    }
  }
  if (markers.length) {
    ast = null;
  }
  monaco.editor.setModelMarkers(model, "owner", markers);
  return {
    ast,
  };
}

export function validate(value: string) {
  const diagnosisNodes: Array<ValidateSchemaGuardMate> = [];
  let ast: Program | null = null;
  try {
    ast = parse.parse(value, {
      ecmaVersion: "latest",
      sourceType: "script",
      locations: true,
    });
    validateWalk(ast as Node, diagnosisNodes);
  } catch (error) {
    if (diagnosisNodes.length === 0) {
      const position = extractTokenAndNumbers((error as Error).message);
      if (position) {
        diagnosisNodes.push(cratedFakeNodeError(position));
      }
    }
    console.log("ignore parse error");
  }
  return {
    diagnosisNodes,
    ast,
  };
}

function cratedFakeNodeError(position: any) {
  return {
    node: {
      loc: {
        start: position,
        end: position,
      },
    },
  } as any;
}
