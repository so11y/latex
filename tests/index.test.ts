import { describe, expect, test } from "vitest";
import { Latex } from "../src/analysis/validate";
import LiteralText from "./text/literal.json";
import { pick } from "lodash-es";
import binaryExpressionText from "./text/binaryExpression.json";
import logicalExpressionText from "./text/logicalExpression.json";
import conditionalExpression from "./text/conditionalExpression.json";

function builderTest(name: string, json: Array<Record<string, any>>) {
  describe(name, () => {
    json.forEach((item, index) => {
      test(`${item.title}__(${index})`, () => {
        const { markers } = Latex.getInstance().getMarkers(item.value);
        expect(markers.length === 0).toEqual(item.expect);
        expect(
          markers.map((marker) => pick(marker, ["message"]))
        ).toMatchSnapshot();
      });
    });
  });
}
builderTest("literal", LiteralText);
builderTest("binaryExpression", binaryExpressionText);
builderTest("logicalExpression", logicalExpressionText);
builderTest("conditionalExpression", conditionalExpression);
