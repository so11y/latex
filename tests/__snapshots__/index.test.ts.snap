// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`binaryExpression > &  不支持的运算符__(5) 1`] = `
[
  {
    "message": "不支持位运算等符号",
  },
]
`;

exports[`binaryExpression > * 号运算__(2) 1`] = `[]`;

exports[`binaryExpression > **  不支持的运算符__(7) 1`] = `
[
  {
    "message": "不支持位运算等符号",
  },
]
`;

exports[`binaryExpression > + 号运算__(0) 1`] = `[]`;

exports[`binaryExpression > / 号运算__(3) 1`] = `[]`;

exports[`binaryExpression > |  不支持的运算符__(6) 1`] = `
[
  {
    "message": "不支持位运算等符号",
  },
]
`;

exports[`binaryExpression > - 号运算__(1) 1`] = `[]`;

exports[`binaryExpression > 1 + '1'__(8) 1`] = `
[
  {
    "message": "运算两边需要是数值或者函数调用表达式或者运算",
  },
]
`;

exports[`binaryExpression > 复杂运算在三元表达式中__(4) 1`] = `[]`;

exports[`conditionalExpression > Conditional语法糖__(3) 1`] = `[]`;

exports[`conditionalExpression > 三元表达式__(0) 1`] = `
[
  {
    "message": "只能是使用三元和逻辑表达式符号 == != > < >= <= && ||",
  },
  {
    "message": "真结果需要是函数调用并且不能再返回逻辑表达式了",
  },
  {
    "message": "假结果需要是函数调用并且不能再返回逻辑表达式了",
  },
]
`;

exports[`conditionalExpression > 三元表达式__(1) 1`] = `
[
  {
    "message": "真结果需要是函数调用并且不能再返回逻辑表达式了",
  },
  {
    "message": "假结果需要是函数调用并且不能再返回逻辑表达式了",
  },
]
`;

exports[`conditionalExpression > 三元表达式__(2) 1`] = `[]`;

exports[`conditionalExpression > 不支持三元表达式嵌套__(4) 1`] = `
[
  {
    "message": "不支持在直接嵌套Conditional函数和三元表达式,但是可以在搭配逻辑表达式使用",
  },
]
`;

exports[`conditionalExpression > 支持三元表达式嵌套搭配逻辑表达式__(5) 1`] = `[]`;

exports[`expressionStatement > 作为root的三元表达__(4) 1`] = `[]`;

exports[`expressionStatement > 作为root的条件表达式-1__(0) 1`] = `[]`;

exports[`expressionStatement > 作为root的条件表达式-2__(1) 1`] = `
[
  {
    "message": "最外层只能是函数调用或者运算或者三元表达在或者是数值",
  },
]
`;

exports[`expressionStatement > 作为root的条件表达式-3__(2) 1`] = `
[
  {
    "message": "最外层只能是函数调用或者运算或者三元表达在或者是数值",
  },
]
`;

exports[`expressionStatement > 作为root的运算-1__(5) 1`] = `[]`;

exports[`expressionStatement > 作为root的逻辑-2__(3) 1`] = `
[
  {
    "message": "当前最终运算得出的可能是一个布尔值，请检查",
  },
]
`;

exports[`literal > 一个字符__(1) 1`] = `
[
  {
    "message": "字符串只能出现在调用里",
  },
]
`;

exports[`literal > 一个数值__(0) 1`] = `[]`;

exports[`literal > 函数中作为字符串__(9) 1`] = `[]`;

exports[`literal > 数值在 !=  逻辑符中__(2) 1`] = `[]`;

exports[`literal > 数值在 *  运算符中__(7) 1`] = `[]`;

exports[`literal > 数值在 +  运算符中__(4) 1`] = `[]`;

exports[`literal > 数值在 /  运算符中__(6) 1`] = `[]`;

exports[`literal > 数值在 -  运算符中__(5) 1`] = `[]`;

exports[`literal > 数值在函数中作为数值__(3) 1`] = `
[
  {
    "message": "参数需要是字符串",
  },
]
`;

exports[`literal > 数值在括号中__(8) 1`] = `[]`;

exports[`logicalExpression > || 号运算 作为Root__(2) 1`] = `
[
  {
    "message": "逻辑表达式两边需要是条件表达式 使用 && 或者 || 连接 ",
  },
]
`;

exports[`logicalExpression > 逻辑运算在三元表达式中__(0) 1`] = `[]`;

exports[`logicalExpression > 逻辑运算在三元表达式中__(1) 1`] = `
[
  {
    "message": "逻辑表达式两边需要是条件表达式 使用 && 或者 || 连接 ",
  },
]
`;
